File: src/10장/10-5-2.py
import sys

q, x = map(int, sys.stdin.readline().split())
map = {}
mex = 0
for _ in range(q):
    y = int(sys.stdin.readline())
    if y % x in map:
        if y % x+x*map[y % x] in map:
            map[y % x+x*map[y % x]] += 1
        else:
            map[y % x+x*map[y % x]] = 1
        map[y % x] += 1
    else: map[y % x] = 1

    while mex in map:
        mex += 1
    print(mex)

File: src/10장/10-5-2.java
import java.util.*;
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int q = Integer.parseInt(st.nextToken());
        int x = Integer.parseInt(st.nextToken());
        int[] map = new int[x];
        for(int i = 0; i < x; i++) {
            map[i] = i;
        }
        boolean[] has = new boolean[q+1];
        int mex = 0;
        for(int i = 0; i < q; i++) {
            int y = Integer.parseInt(br.readLine());
            if(map[y%x] < q) {
                has[map[y%x]] = true;
                while(has[mex]) {
                    mex++;
                }
                map[y%x] += x;
            }
            out.println(mex);
        }
        br.close();
        out.close();
	}
}